### ISSUE TITLE
Order of tokens

### ISSUE THREAD

**Issue by paxxux** [JetBrains/Grammar-Kit#130]
Hi,

I want to parse something like:

`class A inherit B subclass C private`

But the order of tokens and the appearance of: inherit B, subclass C, and private is arbitrary. So if I come up with `"class" id (("inherit" id) | ("subclass" id) | ("private" | "public"))*` the keywords won't be unique, but I need them to be unique and in arbitrary order.

So, as far as I know the only way to this, is with an external rule. I am not capable at my current level of Grammar-Kit knowledge to make this working.

Could you please provide some code on how to achieve this? Is there an "unique" option I missed (would be great if there were something like this)?

Kind regards
paxxux

**Comment by calexHG** [comment]
While I'm sure you could so if you were adamant about finding a way to
implement this in the parser (and, trust me, I understand that that would
be the natural way of thinking about this at first), what you'll probably
find is that the best way to look at this is as a semantic error rather
than as a syntactic error.

In other words, I believe a more common approach would be to have the
parser be partially blind to exact semantic relationship between those
clauses, and to treat whatever code passes your grammar file as acceptable
on the syntactical level.
1. Accept however many "inherit" and "subclass" clauses as you want as
valid SYNTAX.
2. Use IntelliJ' Inspections and Annotators to spot and highlight the cases
where SEMANTICS don't make sense, such as when one of those clauses
overrides or conflicts with a previously typed clause.

Essentially, as your language plugin gets more developed, I believe that
less and less of the errors you will be focusing on will be of the purely
syntactic kind; they will be semantic errors, and they will be best handled
*after* the parser.

(Side note: I'm feeling pretty groggy/tired-minded as I'm writing this,
so.. there's a good chance that I'll wake up tomorrow and edit this to say
something completely different :p)

**Comment by paxxux** [comment]
Hi,

Thank you for your answer. It basically sounds like a good plan to handle this. But I need the parser for the idea plugin and standalone. As far as I know I cannot work with inspections and annotators in the standalone version? So I think I should not allow the parser to pass it?

I want to build a parser for a context-sensitive language. Where for example: 
```
IF = 2
THEN = "hi"
ELSE = 2
IF IF = ELSE THEN SAY THEN --output: hi
```
is valid code. I am sure I have to do some "after-parsing" to fix the AST. Could you tell me how I can achieve this in standalone and plugin with a minimum of adaption? Where would be the best place to do this?

For the "if statement" I did the following in BNF:

`if_stmt ::= IF (bool_expression | ",")+ ";"? THEN ";"? instruction else_stmt?`

this works fine until I allow IF or THEN as identifiers which are part of the boolean expression. 
```
IF = 2
THEN = IF IF + 2 IF IF
SAY THEN -- output: 2 4 2 2
```
is also valid, which make the if statement even worse (
```
IF = 2 
IF IF IF = IF IF THEN SAY "HI" THEN SAY --output: HI THEN SAY
```
to handle it.
Could you give me a clue on how to make the if statement working? (This issue arises everywhere not only with if). By the way, the language is OoRexx.

Best regards

**Comment by calexHG** [comment]
Sorry that I cannot spend time to elaborate on the various design paths
available, the pro's and con's that they may present, and evaluate your
specific project's best design path.

I can offer a few tips tho.

Regarding using the parser for both IDE usage and standalone usage:

IDE requirements are more demanding than, say, a standalone compiler's
requirements, because an IDE has to keep up with an ever-changing program.
With every keystroke the user makes, technically, an entirely different
program may exist, so in that realm it is common to delay much of the
program analysis until later in order to keep the IDE performant. (In the
IDE realm, you also may be interested in allowing slightly improper grammar
so that you can provide more helpful errors that include quick-fixes.) I'll
leave it to you to understand why that distinction may be relevant to you.

As for simply making the language traits you mentioned in your latest
response work, I can list a few notions that may be useful. (What actually
really makes sense for you would depend on how expansive your language is
(ie. how many syntaxes might you need to be cautious of and distinguish),
and also how simple or complex your language is (for example, whether you
can nest control flow into expressions).)

1. Predicates

Predicates can be used to narrow the amount of situations in which a rule
will be used, in order to distinguish between one kind of usage and
another, otherwise-similar kind of usage.

2. Define different rule sets for different contexts or statement depths,
as appropriate for your language.

For example, at the top level, your language might only accept what you
might call a statement. But, within that statement, it may only accept what
you call an expression. And, as in your example, it might be that an
expression can't possibly start with an "if" statement, because it is an
expression and not a statement, which automatically clears up any confusion
as to whether "if if" is an if-statement or an expression.

Basically, your situation is probably not very complicated at all. You just
need to properly define your accepted grammars per different context levels
(ie. have a statement ruleset that differs from an expression ruleset).


And back to your original question, no, there is no "unique" flag in GK.
(Note: if your concern was simply whether "implement" and whatever the
other keyword was only used once each, and I wanted it done at validated at
parse-time, I'd personally find it acceptable to simply hardcode the two
different orders the could come in, since there are only two different
order permutations.)

Toodles.

**Comment by paxxux** [comment]
Thank you very much for your answer! I solved my issues! You helped me a lot!

Best regards

**Comment by calexHG** [comment]
@paxxux That's great! I later wondered whether my answer was too abstract for someone to figure out what I was really trying to say [yet couldn't afford to elaborate further on]. 

...So I'm glad you came back to share the fact that I wasn't wasting my breath — that my insights were able to help you solve your problem. 

Grammar-Kit really is a fantastic tool once you figure out how to use it correctly. Feel free to come on back here if you have any other issues. (Actually, the project maintainer @gregsh might prefer that non-bug-reports/help questions of this sort instead get directed to the IntelliJ API forums, where he and other knowledgeable folks are more present, perhaps..? I'm really not sure.)

**Comment by gregsh** [comment]
As @calexHG I prefer inspections not to complicate grammars. But unique keys can be implemented via external methods + ThreadLocals. We did it once like that:
````
items ::= <<withFlags (item1 | item2 | item3)*>>  // clears some map
item1 ::= <<withFlag 'item1' parse_item1>>           // sets-if-missing flag 'item1' & parse_item1
item2 ::= <<withFlag 'item2' parse_item1>>          // .. same for item2 and item3
item3 ::= <<withFlag 'item3' parse_item1>>          // If a flag is set 'withFlag' returns false
````

PS
Generally it is OK to report/ask by all google'able means. Every venue nowadays has its own pro's and con's. More people view IntelliJ API forums which makes them more appropriate for general how-to-questions; while GK issues here are kept in one place.

**Comment by paxxux** [comment]
I very appreciate your help guys!

Thanks again!
